src/Compilers/Core/CodeAnalysisTest/Diagnostics/CouldHaveMoreSpecificTypeAnalyzer.cs

```diff
@@ -113,7 +113,21 @@ public sealed override void Initialize(AnalysisContext context)
                                 IFieldSymbol containingField = operationBlockContext.OwningSymbol as IFieldSymbol;
                                 if (containingField != null)
                                 {
-                                    AssignTo(containingField, containingField.Type, fieldsSourceTypes, (IExpression)operationBlockContext.OperationBlocks[0]);
+                                    IFieldInitializer initializer = operationBlockContext.OperationBlocks[0] as IFieldInitializer;
+                                    if (initializer != null)
+                                    {
+                                        AssignTo(containingField, containingField.Type, fieldsSourceTypes, initializer.Value);
+                                    }
+                                    else
+                                    {
+                                        // Visual Basic "As New" field declarations don't produce field initializations--they just produce object creation expressions.
+                                        // This is a bug in the VB semantic model support, and requires a subtle fix. For now, code around the issue in this analyzer.
+                                        IExpression expression = operationBlockContext.OperationBlocks[0] as IExpression;
```

Comment:
This code shouldn't need to be here. I have a question out to @AlekseyTs about how best to cope with the semantic model issue.